connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -1247,6 +1259,217 @@ void sourceConnectorOffsets(String connName, ConnectorOffsetBackingStore offsetS
+                timer.update();
+            }
+
+            Set<TopicPartition> partitionsToReset = parsedOffsets.entrySet()
+                    .stream()
+                    .filter(entry -> entry.getValue() == null)
+                    .map(Map.Entry::getKey)
+                    .collect(Collectors.toSet());
+
+            if (!partitionsToReset.isEmpty()) {
+                DeleteConsumerGroupOffsetsResult deleteConsumerGroupOffsetsResult = admin.deleteConsumerGroupOffsets(groupId, partitionsToReset);
+                try {
+                    log.debug("Deleting the consumer group offsets for the following topic partitions using an admin client for sink connector {}: {}.",
+                            connName, partitionsToReset);
+                    deleteConsumerGroupOffsetsResult.all().get(timer.remainingMs(), TimeUnit.MILLISECONDS);
+                } catch (ExecutionException e) {
+                    // TODO: Handle different exception cause types to surface more fine-grained errors here?
+                    throw new ConnectException("Failed to delete consumer group offsets for topic partitions " + partitionsToReset + " for connector "
+                            + connName, e.getCause());
+                } catch (TimeoutException e) {
+                    throw new ConnectException("Timed out while attempting to delete consumer group offsets for topic partitions " + partitionsToReset
+                            + " for connector" + connName, e);
+                } catch (InterruptedException e) {
+                    throw new ConnectException("Unexpectedly interrupted while attempting to delete consumer group offsets for topic partitions " + partitionsToReset
+                            + " for connector" + connName, e);
+                }
+            }
+            return alterOffsetsResult;
+        }
+    }
+
+    /**
+     * Alter a source connector's offsets.
+     * @param connName the name of the source connector whose offsets are to be altered
+     * @param connector an instance of the source connector
+     * @param connectorConfig the source connector's configuration
+     * @param offsets a mapping from partitions to offsets that need to be overwritten
+     * @return true if the source connector has implemented {@link org.apache.kafka.connect.source.SourceConnector#alterOffsets(Map, Map)}
+     * and it returns true for the provided offsets, false otherwise
+     */
+    private boolean alterSourceConnectorOffsets(String connName, Connector connector, Map<String, String> connectorConfig,
+                                                Map<Map<String, ?>, Map<String, ?>> offsets) {
+        SourceConnectorConfig sourceConfig = new SourceConnectorConfig(plugins, connectorConfig, config.topicCreationEnable());
+        Map<String, Object> producerProps = config.exactlyOnceSourceEnabled()
+                ? exactlyOnceSourceTaskProducerConfigs(new ConnectorTaskId(connName, 0), config, sourceConfig,
```

Comment:
It would be cleaner to use a connector specific producer transactional ID here but that would require additional ACLs over what is described in [KIP-618](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors#KIP618:ExactlyOnceSupportforSourceConnectors-Producer) (unless a wildcard prefix is used). @C0urante would it be possible to amend [KIP-875](https://cwiki.apache.org/confluence/display/KAFKA/KIP-875%3A+First-class+offsets+support+in+Kafka+Connect) and document this additional required ACL for altering / resetting offsets for a source connector when exactly once support is enabled on the worker?

Reply:
We could always use the transactional ID for task 0, but that feels a little unclean. I'll ping the vote thread with this and the other updates previously discussed.